ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/ChainDataPruner.java

```diff
@@ -0,0 +1,188 @@
+        BlockchainStorage.Updater updater = blockchainStorage.updater();
+        for (Hash toPrune : oldForkBlocks) {
+          Optional<BlockBody> maybeBody = blockchainStorage.getBlockBody(toPrune);
+          if (maybeBody.isEmpty()) {
+            continue;
+          }
+          // Prune block header, body, receipts, total difficulty and transaction locations.
+          updater.removeBlockHeader(toPrune);
+          updater.removeBlockBody(toPrune);
+          updater.removeTransactionReceipts(toPrune);
+          updater.removeTotalDifficulty(toPrune);
+          maybeBody
+              .get()
+              .getTransactions()
+              .forEach(t -> updater.removeTransactionLocation(t.getHash()));
+        }
+        // Prune canonical chain mapping and commit.
+        updater.removeBlockHash(pruningMark);
+        updater.commit();
+        // Remove old fork blocks.
+        removeForkBlocks(tx, pruningMark);
+        pruningMark++;
+      }
+    }
+    // Update pruning mark and commit
+    setPruningMark(tx, pruningMark);
+    tx.commit();
+  }
+
+  private Optional<Long> getPruningMark() {
+    return get(VARIABLES_PREFIX, PRUNING_MARK_KEY).map(UInt256::fromBytes).map(UInt256::toLong);
+  }
+
+  private Collection<Hash> getForkBlocks(final long blockNumber) {
+    return get(FORK_BLOCKS_PREFIX, UInt256.valueOf(blockNumber))
+        .map(bytes -> RLP.input(bytes).readList(in -> bytesToHash(in.readBytes32())))
+        .orElse(Lists.newArrayList());
+  }
+
+  private void setPruningMark(
+      final KeyValueStorageTransaction transaction, final long pruningMark) {
+    set(transaction, VARIABLES_PREFIX, PRUNING_MARK_KEY, UInt256.valueOf(pruningMark));
+  }
+
+  private void setForkBlocks(
+      final KeyValueStorageTransaction transaction,
+      final long blockNumber,
+      final Collection<Hash> forkBlocks) {
+    set(
+        transaction,
+        FORK_BLOCKS_PREFIX,
+        UInt256.valueOf(blockNumber),
+        RLP.encode(o -> o.writeList(forkBlocks, (val, out) -> out.writeBytes(val))));
+  }
+
+  private void removeForkBlocks(
+      final KeyValueStorageTransaction transaction, final long blockNumber) {
+    remove(transaction, FORK_BLOCKS_PREFIX, UInt256.valueOf(blockNumber));
+  }
+
+  private Optional<Bytes> get(final Bytes prefix, final Bytes key) {
+    return prunerStorage.get(Bytes.concatenate(prefix, key).toArrayUnsafe()).map(Bytes::wrap);
+  }
+
+  private void set(
+      final KeyValueStorageTransaction transaction,
+      final Bytes prefix,
+      final Bytes key,
+      final Bytes value) {
+    transaction.put(Bytes.concatenate(prefix, key).toArrayUnsafe(), value.toArrayUnsafe());
+  }
+
+  private void remove(
+      final KeyValueStorageTransaction transaction, final Bytes prefix, final Bytes key) {
+    transaction.remove(Bytes.concatenate(prefix, key).toArrayUnsafe());
+  }
+
+  private Hash bytesToHash(final Bytes bytes) {
+    return Hash.wrap(Bytes32.wrap(bytes, 0));
+  }
```

Comment:
Would be cleaner to move the pruning storage out into it's own class

Reply:
Added a separate class for the storage